Added a sorting dropdown to the list view

cole-flexsim opened the pull request:
I added a dropdown to the list view UI that will allow the user to define the columns they want to sort by. I added documentation to the ./docs/list-view.md to explain how to configure the sortableBy function. Here are examples of the new UI: 

Default (the sortableBy() function returns an empty array)
![resize without sorting](https://user-images.githubusercontent.com/71522167/119372760-ecccb900-bc74-11eb-956a-a29b271d4cf8.gif)

Default with sortableBy returning an array
![sorting resize](https://user-images.githubusercontent.com/71522167/119372853-066e0080-bc75-11eb-9b6d-92df692fc7fa.gif)

Large Screen Dropdown
![large screen sorting](https://user-images.githubusercontent.com/71522167/119372910-184fa380-bc75-11eb-8be1-d8c5ea0af49a.gif)

Small Screen Dropdown
![small screen sorting](https://user-images.githubusercontent.com/71522167/119372954-21d90b80-bc75-11eb-8fe1-061012f237da.gif)

Gustavinho commented:
Hey @cole-flexsim this is a great feature, thanks!!!!

I have some comments
- Could you please remove the changes about code styles? just to avoid making noise on this PR and review only the changes for this feature 🙏 
- Feel free to use blade components like `<x-component />`, since I started with support for laravel 6 there are some components using `@component()` but I'm moving them to use the other way

Btw this feature can be replicated to the grid view, it would be great if you could add it to it, or we can open another PR with that change after merging this one.

cole-flexsim commented:
@Gustavinho What exactly do you mean about the changes for code styles?

Did you mean the changes I made to some of the tailwind classes in the blade files? I tried to separate out some of the styles unique to the components from styles that determine where they appear on the page

If you are referencing the vscode settings I only added that so that the indentation for someone using the formatter will be 2 spaces just like you already have in your project. My auto formatter defaults to 4 spaces indentation in blade files so I figured it might be helpful to have that setting for anyone else who contributes to the project. Happy to remove it though if it's not something you want in the project

Gustavinho commented:
I mean like this one
![image](https://user-images.githubusercontent.com/4228757/119547044-3c38e500-bd5a-11eb-8be8-a843bc3bd020.png)
![image](https://user-images.githubusercontent.com/4228757/119547308-8fab3300-bd5a-11eb-872f-74ff5bf5a9ba.png)

Those changes are not related to the feature that you implemented, I think your text editor changed them automatically.

Gustavinho commented on resources/views/components/drop-down.blade.php:
```diff
@@ -5,35 +5,29 @@
-  class="text-right relative"
-  x-data="{ open: false }"
->
+<div class="relative" x-data="{ open: false }">
```
Hi @cole-flexsim these are the kind of changes about code style that I was referring to in my comment

Gustavinho replied:
thanks! that way we can focus the PR on the actual feature

cole-flexsim commented:
@Gustavinho  I went through and got rid of most of the formatting changes that happened. I am going to start working on making the sorting dropdown available on the GridView as well as the ListView. Since those two views will be sharing functionality I wanted to get your opinion on how to set that up. 

I could make an abstract SequenceView class that extends DataView. Then ListView and GridView extend SequenceView. 

The other option is I could make a Trait in the package that ListView and GridView use that adds the sortableBy functionality to those classes. If I go this route what would you like the namespace for that trait to be?

Gustavinho commented:
@cole-flexsim I think a Trait is a better approach since there is already one for using actions, `LaravelViews\Actions\WithActions`, I think now it is worth it to have a namespace dedicated for these traits, what about `LaravelViews\Views\Traits`?

Gustavinho commented on src/LaravelViews.php:
```diff
@@ -23,6 +23,8 @@ public function components()
             'alert' => 'alert',
             'alerts-handler' => 'alerts-handler',
+            'form.input-group' => 'form.input-group',
+            'drop-down' => 'drop-down',
```
I think there is already a `drop-down` registered

Gustavinho commented on resources/views/table-view/filters.blade.php:
```diff
@@ -8,50 +8,10 @@
-    @endif
+  {{-- Filters --}}
+  <div class="lg:flex-1 ml-4 lg:ml-0 text-right">
+    @include('laravel-views::components.filters.filters')
```
Nice 👍

Gustavinho commented on resources/views/list-view/sorting.blade.php:
```diff
@@ -0,0 +1,36 @@
+  - form/select --}}
+<div class="leading-tight">
+  {{-- Sorting dropdown --}}
+  @component('laravel-views::components.drop-down')
```
You can use blade components like `<x-lv-drop-down />` instead, I used this`@components` way since I was supporting Laravel 6

Gustavinho commented on resources/views/list-view/list-view.blade.php:
```diff
@@ -1,7 +1,7 @@
   {{-- Search input and filters --}}
-  <div class="px-4">
-    @include('laravel-views::table-view.filters')
+  <div class="pl-4">
```
The padding on both sides was to align it up with the list items, do you think looks better this way?

Gustavinho commented on resources/views/list-view/filters.blade.php:
```diff
@@ -0,0 +1,26 @@
+  </div>
+
+  {{-- Sorting --}}
+  @if (isset($sortableBy) && $sortableBy->isNotEmpty())
```
Not sure if I like how it looks on 'md' or below, what if we pull the sorting button (regular size) to the left and the filters button to the right?

Gustavinho commented on resources/views/list-view/sorting.blade.php:
```diff
@@ -0,0 +1,36 @@
+  @component('laravel-views::components.drop-down')
+    @slot('trigger')
+      @include('laravel-views::components.buttons.select', [
+      'title'=> $sortableBy->flip()->get($sortBy) ?? __('Sort By'),
```
it'd be great if the button title says something like `Sort by: {sortOrder} {field}` or something like that, once you have selected something, it becomes unclear what the button is for, what do you think?

Gustavinho commented on resources/views/components/buttons/select.blade.php:
```diff
@@ -0,0 +1,20 @@
+ - title 
+ - block
+ - variant --}}
+<button
```
I think it would be even better to be able to add an icon to the main button component, the icon could be passed as a prop, what do you think?

Gustavinho commented on resources/views/list-view/filters.blade.php:
```diff
@@ -0,0 +1,26 @@
+  @endif
+
+  {{-- Filters --}}
+  <div class="text-right {{ isset($sortableBy) && $sortableBy->isNotEmpty() ? '' : 'lg:flex-1 ml-4 lg:ml-0' }}">
```
Is this correct?  is the`$sortableBy` conditioning the render of the `filters` component? I think this should be as it was before, although I liked using a collection instead of an array

Gustavinho commented on resources/views/list-view/sorting.blade.php:
```diff
@@ -0,0 +1,36 @@
+      {{ __('Sort By') }}
+    </div>
+    @foreach ($sortableBy as $title => $column)
+      <div class="hover:bg-gray-200">
```
Looks good, either way, you could use the same styles as the `actions.icon-and-title` components, at the end both of them are items (submenu) in a dropdown

Gustavinho reviewed (commented):
Finally I was able to review the code, it looks so good overall, I loved it, I just left a couple of comments, we can discuss them

cole-flexsim commented:
@Gustavinho I just fixed most of the things you mentioned about the code and I added the WithSortableDropdown trait to handle this functionality on the ListView and the GridView

Gustavinho commented:
Great!!!! I'll merge it for the `2.3.0` version

Gustavinho commented:
Hi @cole-flexsim I did some small tweaks on the UI, what do you think?

### Desktop
![image](https://user-images.githubusercontent.com/4228757/120089211-bb614c80-c0bd-11eb-987a-f56e25be0fca.png)
![image](https://user-images.githubusercontent.com/4228757/120089217-c4eab480-c0bd-11eb-99e2-9f06be33a819.png)

### Mobile
![image](https://user-images.githubusercontent.com/4228757/120089221-d9c74800-c0bd-11eb-9f16-3885d7a56318.png)

cole-flexsim commented:
I like it! That looks very clean
